### ISSUE TITLE
Learning Rate Scheduler support

### ISSUE THREAD

**Issue by kparichay** [nnstreamer/nntrainer#1776]
Support for an independent learning rate scheduler interface is needed.
- [x] create learning rate scheduler interface - must be easy to support custom scheduler #1776
- [x] support existing learning rate schedulers (constant, and exponential) with this interface #1776
- [x] register with appcontext for creation by type and later register other schedulers as they come and add unittests #1778
- [x] decouple learning rate for optimizer
  - [x] support new ini section `LearningRateScheduler` with properties `type=constant/exponential`, `learning_rate=1.0`, .... #1783
  - [x] for backward compatibility, the properties and ini interface must support this and pass the request to lr scheduler interface #1782
  - [x] remove learning rate member functions from optimizer but the properties must be supported for backward compatibility #1782
    - [x] create optimizer_scheduler - containing both optimizer and learning rate scheduler - OptimizerWrapped
    - [x] optimizer_scheduler will inherit ml::train::optimizer and will composite nntrainer::optimizer - OptimizerWrapped
- [x] support piecewise constant learning rate type #1784 
- [x] support C++ API #1786
- [x] support C API #2183

**Comment by kparichay** [comment]
For the ini interface/API, there are two options:
```ini
# Option 1
[optimizer]
type = adam
learning_rate_scheduler=cosine
learning_rate=0.1
decay_steps=100

# Option 2
[optimizer]
type = adam
# for backward compatibility, learning_rate, decay_steps and decay_rate properties
# will be supported here, which will internally result in creation of constant/exponential
# learning rate schedule

[LearningRateScheduler]
type=cosine
decay_steps=100
learning_rate=0.1
```

The c/c++ API will look like below:
```c++
// option 1
auto optimizer = ml::train::optimizer::Adam({"learning_rate_scheduler=cosine", "learning_rate=0.1", "decay_steps=100"});

// option 2
auto optimizer = ml::train::optimizer::Adam(); 
auto lr = ml::train::learning_rate_scheduler::Cosine({"learning_rate=0.1", "decay_steps=100"});
optimizer.setLearningRate(lr);
// optimizer.setProperty({"learning_rate=0.1", "decay_steps=100"}); will be available for constant/exponential learning rate as an alternative to optimizer.setLearningRate() for backward compatility
```

I recommend option 2 as its more clearer and decouples the learning-rate scheduler from optimizer details.
Please share your opinions.

**Comment by jijoongmoon** [comment]
I think option 2 is clear and looks better. If it is.. then, we need to c api as well. but it can be done later. Just leave the comment somewhere about c API as well including deprecated notice.

**Comment by zhoonit** [comment]
auto lr = ml::train::learning_rate_scheduler::Cosine({"learning_rate=0.1", "decay_steps=100"}); seems good.

```optimizer.setLearningRate(lr);``` is also good, but please note that this cannot be supported through ini saving protocol, so we need to implement.

**Comment by kparichay** [comment]
> optimizer.setLearningRate(lr); is also good, but please note that this cannot be supported through ini saving protocol, so we need to implement.

The ini saving will serialize the learning rate scheduler object from the optimizer and create the corresponding section in the ini file. In my understanding, it should be just an extension to how the optimizer section will be created in the ini. Let me know if I misunderstood your comment.

**Comment by zhoonit** [comment]
I believe that's about it as well, just noting because easy to miss that part :)
